python/ray/function_manager.py

```diff
@@ -0,0 +1,495 @@
+            # Add the function to the function table.
+            self._worker.redis_client.rpush(
+                b"FunctionTable:" + function_id.id(), self._worker.worker_id)
+
+    def get_execution_info(self, driver_id, function_id):
+        """Get the FunctionExecutionInfo of a remote function.
+
+        Args:
+            driver_id: ID of the driver that the function belongs to.
+            function_id: ID of the function to get.
+
+        Returns:
+            A FunctionExecutionInfo object.
+        """
+        # Wait until the function to be executed has actually been registered
+        # on this worker. We will push warnings to the user if we spend too
+        # long in this loop.
+        with profiling.profile("wait_for_function", worker=self._worker):
+            self._wait_for_function(function_id, driver_id)
+        return self._function_execution_info[driver_id][function_id.id()]
+
+    def _wait_for_function(self, function_id, driver_id, timeout=10):
+        """Wait until the function to be executed is present on this worker.
+
+        This method will simply loop until the import thread has imported the
+        relevant function. If we spend too long in this loop, that may indicate
+        a problem somewhere and we will push an error message to the user.
+
+        If this worker is an actor, then this will wait until the actor has
+        been defined.
+
+        Args:
+            function_id (str): The ID of the function that we want to execute.
+            driver_id (str): The ID of the driver to push the error message to
+                if this times out.
+        """
+        start_time = time.time()
+        # Only send the warning once.
+        warning_sent = False
+        while True:
+            with self._worker.lock:
+                if (self._worker.actor_id == ray.worker.NIL_ACTOR_ID
+                        and (function_id.id() in
+                             self._function_execution_info[driver_id])):
+                    break
+                elif self._worker.actor_id != ray.worker.NIL_ACTOR_ID and (
+                        self._worker.actor_id in self._worker.actors):
+                    break
+                if time.time() - start_time > timeout:
+                    warning_message = ("This worker was asked to execute a "
+                                       "function that it does not have "
+                                       "registered. You may have to restart "
+                                       "Ray.")
+                    if not warning_sent:
+                        ray.utils.push_error_to_driver(
+                            self._worker,
+                            ray_constants.WAIT_FOR_FUNCTION_PUSH_ERROR,
+                            warning_message,
+                            driver_id=driver_id)
+                    warning_sent = True
+            time.sleep(0.001)
+
+    @classmethod
+    def is_function_or_method(cls, x):
```

Comment:
Could we put `is_function_or_method` &  `is_classmethod` into `ray.utils`? They are more like general type inspecting tools.